pps: Use sysfs module parameter to replace PPS_MAX_SOURCES.

SparkingSpork opened the pull request:
Currently, the PPS_MAX_SOURCES is a hard coded constant. On some systems, the current maximum of 16 is too low. This change converts that logic to use a sysfs parameter instead of a hard-coded value and increases the default value to 32.

## Testing:
- Built and installed new kernel on target. 
   - Confirmed that new `/sysfs/module/pps_core/paramters/pps_max_sources` existed and was set to default of 32. 
   - Confirmed value could be changed.
   - Confirmed invalid values errored.
   - Confirmed that adding `pps_core.pps_max_sources=64` to the boot parameters resulted in a value of 64.
- Ran `checkpatch.pl`

Commit:
pps: Use sysfs module parameter to replace PPS_MAX_SOURCES.

Currently, the PPS_MAX_SOURCES is a hard coded constant.
On some systems, the current maximum of 16 is too low.
This change converts that logic to use a sysfs parameter
instead of a hard-coded value.

Signed-off-by: Charlie Johnston <charlie.johnston@ni.com>

SparkingSpork commented on include/linux/pps_kernel.h:
```diff
@@ -67,6 +67,7 @@ struct pps_device {
  */
 
 extern const struct attribute_group *pps_groups[];
+extern int pps_max_sources;
```
Thoughts on `max_sources` instead of `pps_max_sources` since it's a sysfs parameter now?

chaitu236 replied:
Can you clarify what you mean by `since it's a sysfs parameter now`?
Since this is a global variable, I prefer `pps_max_sources` for more specificity.

SparkingSpork replied:
My thought was that it will only be referred to with the context of pps. Previously, code might have `PPS_MAX_SOURCES` in it with no other context. Making it a module parameter means that `pps_core` will typically be part of the calling convention indicating it's part of pps. That being said, I think the more specificity is good since it's used in pps without context so I'll leave it as-is.

SparkingSpork commented on include/uapi/linux/pps.h:
```diff
@@ -26,7 +26,7 @@
 
 #define PPS_VERSION		"5.3.6"
-#define PPS_MAX_SOURCES		16		/* should be enough... */
+#define PPS_MAX_SOURCES		16		/* deprecated, use sysfs instead */
```
Is a `#pragma deprecated(PPS_MAX_SOURCES)` valuable here?

bstreiff replied:
I don't see use of that `#pragma` anywhere else in the uapi headers, and [Linus seems against a `__deprecated` tag](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=771c035372a0). I think annotating it with a comment is the correct thing to do.

SparkingSpork replied:
Cool. Dropping this comment then.

bstreiff commented on drivers/pps/sysfs.c:
```diff
@@ -97,3 +98,28 @@ const struct attribute_group *pps_groups[] = {
+ * Module parameters
+ */
+
+int pps_max_sources = 32;
```
This should probably default to the previous limit of 16; we probably shouldn't change the defaults out from other users.

bstreiff commented on drivers/pps/sysfs.c:
```diff
@@ -97,3 +98,28 @@ const struct attribute_group *pps_groups[] = {
+
+int pps_max_sources = 32;
+
+static int pps_max_sources_write(const char *val, const struct kernel_param *kp)
```
why not use param_set_int here?

Or for that matter, make the parameter a uint and use param_get_uint/param_set_uint, since it's not meaningful to have a negative number of pps sources.

SparkingSpork replied:
When I looked at the function prototype for one of the places it was used, the function input was an `int`. Looking at the other 2 places though, that's not consistent and the other two accept `unsigned`. I'll switch to `uint`.

SparkingSpork replied:
That being said, I'm also realizing this is used to allocate a character device at module init. So I'll still need a function to handle reallocating when it changes which is quite a bit more complicated.

SparkingSpork replied:
Switched to use `module_param()` with a type of uint since the callback isn't needed.

chaitu236 commented on drivers/pps/sysfs.c:
```diff
@@ -97,3 +98,28 @@ const struct attribute_group *pps_groups[] = {
+	.get = param_get_int
+};
+
+module_param_cb(pps_max_sources, &pps_max_sources_ops, &pps_max_sources, 0664);
```
Adding `MODULE_PARM_DESC` may be useful.

SparkingSpork commented:
## V2
- Value is now an unsigned integer.
- Switched to `module_param` instead of `module_param_cb`
- Added `MODULE_PARM_DESC`
- Kept old default maximum of 16

jeminor reviewed (approved):
I will direct message you with a validation request.

bstreiff commented:
Closing in favor of the solution in #131
